### PR TITLE
Full Text Search Support

### PR DESCRIPTION
**Pull request by dplewis**
PR for #3747

Adds Full Text Support for mongo and postgres

### CODE REVIEW

**Comment by seriph** [comment]
Does this support the textScore from MongoDB? I don't see a test for it or any of the code, though everything else is pretty similar to my PR

**Comment by dplewis** [comment]
Working on a test case now. For it unless you want to do it.

**Comment by dplewis** [comment]
To sort by the text score, include the same $meta expression in both the projection document and the sort expression.

Its not as easy as it seems to test with rest.

**Comment by flovilmart** [comment]
Yeah try to coordinate into a single PR if possible @dplewis, you seem well versed into Postgres, why don't you add support to it on @seriph's PR?

**Comment by dplewis** [comment]
@flovilmart I would do that but I don't like how @seriph  copy and pasted my PR an hour after I posted it. And his 'textScore' doesn't work.

He had over a month to do a PR.

Also I announced that I was going to do a PR.

**Comment by flovilmart** [comment]
@seriph's code is way old, and that's open source, let's work together to be more efficient, it's not a time for disputes guys.

**Comment by dplewis** [comment]
@flovilmart how should we proceed? Should I help this first time contributor?

**Comment by flovilmart** [comment]
You gotta treat people with respect man, the way you're acting and telling @seriph down is not acceptable, not matter how many commits you have in the project. @seriph had a fork with the feature in months ago, and I expect he's now pissed and not likely to contribute to the project again. Read our [CODE_OF_CONDUCT](https://github.com/parse-community/parse-server/blob/master/CODE_OF_CONDUCT.md). I can not tolerate your behaviour with other contributors.

**Comment by dplewis** [comment]
I understand it won't happen again.

**Comment by flovilmart** [comment]
Also, on the code side, $text should not be a special key, but the REST format should be:
```
{
  "myKey": {
     "$text": {
        "$search": ...
      }
  }
}
```

**Comment by seriph** [comment]
Haha, to tell you the truth @dplewis I suspected you did the same when I saw you'd posted in the thread where someone had asked me about my fork, then your PR shows up with almost identical code. The textScore code has been live in production on an ecommerce site for months, so surprised you're not thinking it works. Thanks for the professionalism @flovilmart, happy to contribute in the future. I have matching code for the JS API if it helps.

**Comment by flovilmart** [comment]
So we're all good then, if you guys can coordinate, and work together to get the best of that PR, I'll happily create a new full-text-support branch, where you can both merge your PR's before we merge the full-text-support in master. What do you think?

**Comment by dplewis** [comment]
@flovilmart since you can only have 1 index "myKey" doesn't matter.

@seriph I just added a commit for textScore and a test for it. Look at it and you'll see why I thought your textScore didn't work. I didn't see a keyword 'textScore' in any of the docs.

@flovilmart thanks for creating that branch.

**Comment by dplewis** [comment]
@flovilmart do we commit to that branch and move the conversation over to @seriph PR.

**Comment by seriph** [comment]
Happy either way

**Comment by flovilmart** [comment]
Is that true as well for Postgres? (The single index?)

**Comment by dplewis** [comment]
Postgres doesn't care, but why use full text search without an index

**Comment by dplewis** [comment]
Mongo throws an error if index doesn't exist. I'm writing a test case for it now.

**Comment by flovilmart** [comment]
And on Postgres, there's no need either to specify which key of the table will be looked up for search?

**Comment by dplewis** [comment]
I'm still doing research on it. But we'll know soon enough.

But from the looks of it were going to use SELECT [mykey] FROM [className] so maybe.

**Comment by dplewis** [comment]
I still have 4 test cases to write then we can merge.

Index Exists
Multiple Index Error
Space Delimiter
Phrases

**Comment by SebC99** [comment]
@dplewis are you sure you need to to change the `find` function of the `mongoStorageAdapter`? @seriph's version with an update of `mongoCollection` seems simpler, but I'm not good enough in JS to judge :)
I just feel adding those 16 lines in `mongoStorageAdapter` is less clean than the rest of the code :)
My 2 cents

**Comment by dplewis** [comment]
For some reason I couldn't get Mongocollection.find(query...) to handle 2 projections `mongoWhere, {{score...}}`. It could be written but I spent too much time on it and didn't want to break anything existing. Any better solutions would be appreciated.

**Comment by dplewis** [comment]
@SebC99 `findOperation = findOperation.project(keys);` was the line i was looking for to work with projections.

@seriph I wrote a test to work with your solution.

This works for me
```score = {$meta: 'textScore'};```
This doesn't and returns undefined.
```textScore = {$meta: 'textScore'}```

Do you know why?

**Comment by dplewis** [comment]
@flovilmart I think this is good to go for mongo. Unless more tests are needed. I'm going to be out of town for a few days and I'll do postgres when I come back.

**Comment by dplewis** [comment]
@flovilmart I know theres it_exclude_dbs(['postgres']) but is there a if (db == 'postgres') or should I just  rewrite the test for postgres

**Comment by dplewis** [comment]
@flovilmart I've added support for Postgres. Users have to keep in mind that this will have unexpected consequences if their indexes aren't set up properly

**Comment by flovilmart** [comment]
closing to force travis to pick up the build

File: spec/ParseQuery.spec.js [hunk 1]
```diff
@@ -2816,4 +2821,363 @@ describe('Parse.Query testing', () => {
...
+          'X-Parse-REST-API-Key': 'test'
+        }
+      });
+    }).then((resp) => {
+      expect(resp.results.length).toBe(2);
+      done();
+    }, done.fail);
+  });
+
+  it_exclude_dbs(['postgres'])('fullTextSearch: $caseSensitive', (done) => {
```
Comment (flovilmart) [inline spec/ParseQuery.spec.js]: why still excluding PG?

Comment (flovilmart) [inline spec/ParseQuery.spec.js]: Or can you group together the tests related to mongo (excluding PG) in a sub describe?

**Comment by flovilmart** [comment]
It's looking super great IMHO, @seriph what do you think?

File: spec/ParseQuery.spec.js (hunk 1 above)
Comment (dplewis) [inline spec/ParseQuery.spec.js]: There are 3 excludes here. I've added comments in the postgres adapter for $caseSensitive and $diraticSensitive.

No Index... You don't need index to do PG
Case Sensitive... Full text doesn't support case sensitive on PG. You could use regex or make a second column for lower case string.
Diratic Sensitive... Can be automated on a per table / class basis based on setup and index on PG.

Should I replace the comments with error instructions in `postgresAdapter` or leave it be?

Comment (flovilmart) [inline spec/ParseQuery.spec.js]: Probably you should Error the query and point to instructions / have a clear way to resolution

**Comment by dplewis** [comment]
I separated the tests and added the error messages

**Comment by flovilmart** [comment]
Sorry to ask for something more, could we put all the tests in a new file? The ParseQuery.spec.js is already huge :) perhaps ParseQuery.FullTextSearch.spec.js?

**Comment by dplewis** [comment]
Sure no problem glad I could help. Where do I update the docs?

**Comment by dplewis** [comment]
@seriph  can you work on the javascript sdk?

**Comment by flovilmart** [comment]
For now let's start with the REST docs, we'll see the SDK's after right?

**Comment by dplewis** [comment]
Right

**Comment by srameshr** [comment]
Any chance that an API for this will come into the JS SDK?

**Pull request merged**